### ISSUE TITLE
Use TCP Fast Open in server.c and possibly update client side TFO implementation?

### ISSUE THREAD

**Issue by candrews** [getdnsapi/getdns#422]
TCP Fast Open eliminates a round trip for TCP connections. Since stubby is performance sensitive and makes many TCP connections to the DNS-over-TLS server, using TCP Fast Open would be a nice improvement. See https://lwn.net/Articles/508865/ for background.

On the client side, it's as simple as setting the `TCP_FASTOPEN_CONNECT` option on the socket.

On the server side, stubby would do something like this on the listening socket:
```
int qlen = 5;
setsockopt(fd, SOL_TCP, TCP_FASTOPEN, &qlen, sizeof(qlen));
```

Chrome and Firefox have supported TCP Fast Open for clients for over a year, and other DNS servers (ex unbound) use it for client and sever connections too.

**Comment by saradickinson** [comment]
@wtoorop @candrews Just played with the 'Transfer issue' beta feature in GitHub to move this to the getdns repo.... :-)

**Comment by odkrys** [comment]
Is TFO only available with GETDNS_TRANSPORT_TCP ? 
I can't see any TFO flag when I set it to GETDNS_TRANSPORT_TLS.

**Comment by odkrys** [comment]
I see. Unbound TFO implement only support for tcp not tls.
And openssl library doesn't looks like supporting TFO yet.

So I tried to compile with gnutls in develop branch but it required gnutls-dane.
gnutls-dane requires libunbound and unbound require libopenssl.

Huh..?

**Comment by wtoorop** [comment]
@odkrys I'm about to do a commit using TCP_FASTOPEN_CONNECT option, which will do TFO for TLS on Linux (I tested 8.8.8.8 and it seems to work).
And yes the circular dependencies are a bit crazy... Also, even if you compile getdns with GnuTLS, we still need OpenSSL for Zero configuration DNSSEC, because of the S/MIME verification of trust-anchors.xml which we cannot do with GnuTLS (yet).

**Comment by odkrys** [comment]
sadly, it needs kernel 4.11+...

**Comment by wtoorop** [comment]
Fixed in 5.2.2
